File: skills/venera-comic-source-builder/lib/builder.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const { retry } = require('./retry');

class ComicSourceBuilder {
    constructor(options) {
        this.options = options;
        this.testResults = {};
        this.report = [];
    }

    async checkDependencies() {
        this.report.push({ type: 'info', message: '检查Python依赖...' });
        
        return new Promise((resolve, reject) => {
            const checker = path.join(__dirname, '../scripts/dependency_checker.py');
            exec(`python3 "${checker}"`, (error, stdout, stderr) => {
                if (error) {
                    const missing = stdout.trim().replace('missing: ', '');
                    this.report.push({ 
                        type: 'warning', 
                        message: `缺少依赖包: ${missing}` 
                    });
                    reject(new Error(`缺少Python依赖，请运行: pip install ${missing}`));
                } else {
                    this.report.push({ type: 'success', message: 'Python依赖检查通过' });
                    resolve();
                }
            });
        });
    }

    async runTests() {
        this.report.push({ type: 'info', message: '\n开始运行测试...' });

        const pythonScript = path.join(__dirname, '../scripts/comic_source_tester.py');
        const args = [
            '--url', this.options.url,
            '--name', this.options.name,
            '--key', this.options.key,
            '--version', this.options.version,
            '--need-login', this.options.needLogin ? 'true' : 'false',
            '--backup-domains', this.options.backupDomains.join(',')
        ];

        return retry(() => {
            return new Promise((resolve, reject) => {
                const child = exec(`python3 "${pythonScript}" ${args.join(' ')}`, (error, stdout, stderr) => {
                    if (error) {
                        reject(new Error(`测试失败: ${error.message}`));
                    } else {
                        try {
                            this.testResults = JSON.parse(stdout);
                            resolve();
                        } catch {
                            reject(new Error('解析测试结果失败'));
                        }
                    }
                });

                child.stdout.on('data', (data) => {
                    const lines = data.toString().split('\n');
                    lines.forEach(line => {
                        if (line.startsWith('[INFO]')) {
                            this.report.push({ type: 'info', message: line.slice(6) });
                        } else if (line.startsWith('[SUCCESS]')) {
                            this.report.push({ type: 'success', message: line.slice(10) });
                        } else if (line.startsWith('[WARNING]')) {
                            this.report.push({ type: 'warning', message: line.slice(10) });
                        } else if (line.startsWith('[ERROR]')) {
                            this.report.push({ type: 'error', message: line.slice(8) });
                        }
                    });
                });
            });
        }, {
            maxRetries: 3,
            delayMs: 3000,
            onRetry: (attempt, max, error) => {
                this.report.push({ 
                    type: 'warning', 
                    message: `测试失败，第 ${attempt}/${max} 次重试: ${error.message}` 
                });
            }
        });
    }

    async generateSource() {
        this.report.push({ type: 'info', message: '\n开始生成漫画源文件...' });

        const template = this.loadTemplate();
        let sourceCode = template;

        sourceCode = sourceCode.replace(/{{NAME}}/g, this.options.name);
        sourceCode = sourceCode.replace(/{{KEY}}/g, this.options.key);
        sourceCode = sourceCode.replace(/{{VERSION}}/g, this.options.version);
        sourceCode = sourceCode.replace(/{{URL}}/g, this.options.url);

        sourceCode = this.fillExploreSection(sourceCode);
        sourceCode = this.fillSearchSection(sourceCode);
        sourceCode = this.fillCategorySection(sourceCode);
        sourceCode = this.fillComicSection(sourceCode);

        const outputPath = path.join(process.cwd(), `${this.options.key}.js`);
        fs.writeFileSync(outputPath, sourceCode);

        this.report.push({ type: 'success', message: `漫画源文件已生成: ${outputPath}` });
        this.outputPath = outputPath;
    }

    loadTemplate() {
        return fs.readFileSync(path.join(__dirname, '../references/template.js'), 'utf8');
    }

    fillExploreSection(sourceCode) {
        if (!this.testResults.explore) return sourceCode;

        const exploreConfig = this.testResults.explore;
        let exploreCode = `    explore = [\n`;

        exploreConfig.sections.forEach(section => {
            exploreCode += `        {\n`;
            exploreCode += `            title: "${section.title}",\n`;
            exploreCode += `            type: "${section.type || 'multiPartPage'}",\n`;
            exploreCode += `            load: async (page) => {\n`;
            exploreCode += `                let res = await Network.get("${section.url}")\n`;
            exploreCode += `                if (res.status !== 200) throw "请求失败"\n`;
            exploreCode += `                let document = new HtmlDocument(res.body)\n`;
            exploreCode += `                let comics = []\n`;
            exploreCode += `                document.querySelectorAll("${section.selector}").forEach(el => {\n`;
            exploreCode += `                    comics.push(new Comic({\n`;
            exploreCode += `                        id: el.getAttribute("${section.idAttr}"),\n`;
            exploreCode += `                        title: el.querySelector("${section.titleSelector}")?.text || "",\n`;
            exploreCode += `                        cover: el.querySelector("${section.coverSelector}")?.getAttribute("${section.coverAttr}") || "",\n`;
            exploreCode += `                    }))\n`;
            exploreCode += `                })\n`;
            exploreCode += `                return { "${section.key}": comics }\n`;
            exploreCode += `            }\n`;
            exploreCode += `        }\n`;
        });

        exploreCode += `    ]\n`;

        return sourceCode.replace(/\{\{EXPLORE_SECTION\}\}/g, exploreCode);
    }

    fillSearchSection(sourceCode) {
        if (!this.testResults.search) return sourceCode;

        const searchConfig = this.testResults.search;
        const searchCode = `    search = {\n` +
            `        load: async (keyword, options, page) => {\n` +
            `            let url = "${searchConfig.url}".replace("{{keyword}}", encodeURIComponent(keyword))\n` +
            `            let res = await Network.get(url)\n` +
            `            if (res.status !== 200) throw "请求失败"\n` +
            `            let document = new HtmlDocument(res.body)\n` +
            `            let comics = []\n` +
            `            document.querySelectorAll("${searchConfig.selector}").forEach(el => {\n` +
            `                comics.push(new Comic({\n` +
            `                    id: el.getAttribute("${searchConfig.idAttr}"),\n` +
            `                    title: el.querySelector("${searchConfig.titleSelector}")?.text || "",\n` +
            `                    cover: el.querySelector("${searchConfig.coverSelector}")?.getAttribute("${searchConfig.coverAttr}") || "",\n` +
            `                }))\n` +
            `            })\n` +
            `            return { comics: comics, maxPage: 10 }\n` +
            `        }\n` +
            `    }\n`;

        return sourceCode.replace(/\{\{SEARCH_SECTION\}\}/g, searchCode);
    }

    fillCategorySection(sourceCode) {
        if (!this.testResults.category) return sourceCode;

        const categoryConfig = this.testResults.category;
        let categoryCode = `    category = {\n` +
            `        title: "分类",\n` +
            `        parts: [\n`;

        categoryConfig.categories.forEach(cat => {
            categoryCode += `            {\n` +
                `                name: "${cat.name}",\n` +
                `                type: "fixed",\n` +
                `                categories: [\n`;
            cat.items.forEach(item => {
                categoryCode += `                    {\n` +
                    `                        label: "${item.label}",\n` +
                    `                        target: {\n` +
                    `                            page: "category",\n` +
                    `                            attributes: { category: "${item.value}" }\n` +
                    `                        }\n` +
                    `                    },\n`;
            });
            categoryCode += `                ]\n` +
                `            },\n`;
        });

        categoryCode += `        ]\n` +
            `    }\n`;

        return sourceCode.replace(/\{\{CATEGORY_SECTION\}\}/g, categoryCode);
    }

    fillComicSection(sourceCode) {
        if (!this.testResults.comic) return sourceCode;

        const comicConfig = this.testResults.comic;
        const comicCode = `    comic = {\n` +
            `        loadInfo: async (id) => {\n` +
            `            let url = "${comicConfig.infoUrl}".replace("{{id}}", id)\n` +
            `            let res = await Network.get(url)\n` +
            `            if (res.status !== 200) throw "请求失败"\n` +
            `            let document = new HtmlDocument(res.body)\n` +
            `            let title = document.querySelector("${comicConfig.titleSelector}")?.text || ""\n` +
            `            let cover = document.querySelector("${comicConfig.coverSelector}")?.getAttribute("${comicConfig.coverAttr}") || ""\n` +
            `            let author = document.querySelector("${comicConfig.authorSelector}")?.text || ""\n` +
            `            let description = document.querySelector("${comicConfig.descSelector}")?.text || ""\n` +
            `            let chapters = []\n` +
            `            document.querySelectorAll("${comicConfig.chapterSelector}").forEach(el => {\n` +
            `                chapters.push(new Chapter({\n` +
            `                    id: el.getAttribute("${comicConfig.chapterIdAttr}"),\n` +
            `                    title: el.querySelector("${comicConfig.chapterTitleSelector}")?.text || "",\n` +
            `                }))\n` +
            `            })\n` +
            `            return new ComicDetails({\n` +
            `                id: id,\n` +
            `                title: title,\n` +
            `                cover: cover,\n` +
            `                author: author,\n` +
            `                description: description,\n` +
            `                chapters: chapters.reverse()\n` +
            `            })\n` +
            `        },\n` +
            `        loadEp: async (comicId, epId) => {\n` +
            `            let url = "${comicConfig.epUrl}".replace("{{comicId}}", comicId).replace("{{epId}}", epId)\n` +
            `            let res = await Network.get(url)\n` +
            `            if (res.status !== 200) throw "请求失败"\n` +
            `            let document = new HtmlDocument(res.body)\n` +
            `            let images = []\n` +
            `            document.querySelectorAll("${comicConfig.imageSelector}").forEach(el => {\n` +
            `                let src = el.getAttribute("${comicConfig.imageAttr}")\n` +
            `                if (src) images.push(src)\n` +
            `            })\n` +
            `            return { images: images }\n` +
            `        }\n` +
            `    }\n`;

        return sourceCode.replace(/\{\{COMIC_SECTION\}\}/g, comicCode);
    }

    showReport() {
        console.log('\n' + '='.repeat(60));
        console.log('          漫画源生成报告');
        console.log('='.repeat(60));

        this.report.forEach(item => {
            let prefix = '  ';
            switch (item.type) {
                case 'success': prefix = '  ✅ '; break;
                case 'error': prefix = '  ❌ '; break;
                case 'warning': prefix = '  ⚠️ '; break;
                case 'info': prefix = '  ℹ️ '; break;
            }
            console.log(prefix + item.message);
        });

        console.log('='.repeat(60));
        if (this.outputPath) {
            console.log(`\n生成的源文件: ${this.outputPath}`);
        }
    }

    async build() {
        try {
            await this.checkDependencies();
            await this.runTests();
            await this.generateSource();
            this.showReport();
            return { success: true, outputPath: this.outputPath };
        } catch (error) {
            this.report.push({ type: 'error', message: `构建失败: ${error.message}` });
            this.showReport();
            throw error;
        }
    }
}

module.exports = ComicSourceBuilder;

File: skills/venera-comic-source-builder/builder.py
import os
import re
import subprocess
import json
from retry import retry

class ComicSourceBuilder:
    def __init__(self, options):
        self.options = options
        self.test_results = {}
        self.report = []
    
    def log(self, level, message):
        self.report.append({'type': level, 'message': message})
    
    async def check_dependencies(self):
        self.log('info', '检查Python依赖...')
        
        checker_path = os.path.join(os.path.dirname(__file__), 'scripts', 'dependency_checker.py')
        result = subprocess.run(['python3', checker_path], capture_output=True, text=True)
        
        if result.returncode != 0:
            missing = result.stdout.strip().replace('missing: ', '')
            self.log('warning', f'缺少依赖包: {missing}')
            raise RuntimeError(f'缺少Python依赖，请运行: pip install {missing}')
        else:
            self.log('success', 'Python依赖检查通过')
    
    async def run_tests(self):
        self.log('info', '\n开始运行测试...')
        
        python_script = os.path.join(os.path.dirname(__file__), 'scripts', 'comic_source_tester.py')
        args = [
            '--url', self.options['url'],
            '--name', self.options['name'],
            '--key', self.options['key'],
            '--version', self.options['version'],
            '--need-login', 'true' if self.options['need_login'] else 'false',
            '--backup-domains', ','.join(self.options['backup_domains'])
        ]
        
        async def run_test_once():
            result = subprocess.run(['python3', python_script] + args, capture_output=True, text=True)
            
            if result.returncode != 0:
                raise RuntimeError(f'测试失败: {result.stderr}')
            
            try:
                return json.loads(result.stdout)
            except json.JSONDecodeError:
                raise RuntimeError('解析测试结果失败')
        
        def on_retry(attempt, max_retry, error):
            self.log('warning', f'测试失败，第 {attempt}/{max_retry} 次重试: {error}')
        
        self.test_results = await retry(run_test_once, max_retries=3, delay_ms=3000, on_retry=on_retry)
    
    def generate_source(self):
        self.log('info', '\n开始生成漫画源文件...')
        
        template_path = os.path.join(os.path.dirname(__file__), 'references', 'template.js')
        with open(template_path, 'r', encoding='utf-8') as f:
            template = f.read()
        
        source_code = template
        source_code = source_code.replace('{{NAME}}', self.options['name'])
        source_code = source_code.replace('{{KEY}}', self.options['key'])
        source_code = source_code.replace('{{VERSION}}', self.options['version'])
        source_code = source_code.replace('{{URL}}', self.options['url'])
        
        source_code = self.fill_explore_section(source_code)
        source_code = self.fill_search_section(source_code)
        source_code = self.fill_category_section(source_code)
        source_code = self.fill_comic_section(source_code)
        
        output_path = os.path.join(os.getcwd(), f"{self.options['key']}.js")
        with open(output_path, 'w', encoding='utf-8') as f:
            f.write(source_code)
        
        self.log('success', f'漫画源文件已生成: {output_path}')
        self.output_path = output_path
    
    def fill_explore_section(self, source_code):
        if not self.test_results.get('explore'):
            return source_code
        
        explore_config = self.test_results['explore']
        explore_code = '    explore = [\n'
        
        for section in explore_config['sections']:
            explore_code += f'''        {{
            title: "{section['title']}",
            type: "{section.get('type', 'multiPartPage')}",
            load: async (page) => {{
                let res = await Network.get("{section['url']}")
                if (res.status !== 200) throw "请求失败"
                let document = new HtmlDocument(res.body)
                let comics = []
                document.querySelectorAll("{section['selector']}").forEach(el => {{
                    comics.push(new Comic({{
                        id: el.getAttribute("{section['idAttr']}"),
                        title: el.querySelector("{section['titleSelector']}")?.text || "",
                        cover: el.querySelector("{section['coverSelector']}")?.getAttribute("{section['coverAttr']}") || "",
                    }}))
                }})
                return {{ "{section['key']}": comics }}
            }}
        }}\n'''
        
        explore_code += '    ]\n'
        return source_code.replace('{{EXPLORE_SECTION}}', explore_code)
    
    def fill_search_section(self, source_code):
        if not self.test_results.get('search'):
            return source_code
        
        search_config = self.test_results['search']
        search_code = f'''    search = {{
        load: async (keyword, options, page) => {{
            let url = "{search_config['url']}".replace("{{keyword}}", encodeURIComponent(keyword))
            let res = await Network.get(url)
            if (res.status !== 200) throw "请求失败"
            let document = new HtmlDocument(res.body)
            let comics = []
            document.querySelectorAll("{search_config['selector']}").forEach(el => {{
                comics.push(new Comic({{
                    id: el.getAttribute("{search_config['idAttr']}"),
                    title: el.querySelector("{search_config['titleSelector']}")?.text || "",
                    cover: el.querySelector("{search_config['coverSelector']}")?.getAttribute("{search_config['coverAttr']}") || "",
                }}))
            }})
            return {{ comics: comics, maxPage: 10 }}
        }}
    }}\n'''
        
        return source_code.replace('{{SEARCH_SECTION}}', search_code)
    
    def fill_category_section(self, source_code):
        if not self.test_results.get('category'):
            return source_code
        
        category_config = self.test_results['category']
        category_code = '    category = {\n        title: "分类",\n        parts: [\n'
        
        for cat in category_config['categories']:
            category_code += f'''            {{
                name: "{cat['name']}",
                type: "fixed",
                categories: [
'''
            for item in cat['items']:
                category_code += f'''                    {{
                        label: "{item['label']}",
                        target: {{
                            page: "category",
                            attributes: {{ category: "{item['value']}" }}
                        }}
                    }},
'''
            category_code += '                ]\n            },\n'
        
        category_code += '        ]\n    }\n'
        return source_code.replace('{{CATEGORY_SECTION}}', category_code)
    
    def fill_comic_section(self, source_code):
        if not self.test_results.get('comic'):
            return source_code
        
        comic_config = self.test_results['comic']
        comic_code = f'''    comic = {{
        loadInfo: async (id) => {{
            let url = "{comic_config['infoUrl']}".replace("{{id}}", id)
            let res = await Network.get(url)
            if (res.status !== 200) throw "请求失败"
            let document = new HtmlDocument(res.body)
            let title = document.querySelector("{comic_config['titleSelector']}")?.text || ""
            let cover = document.querySelector("{comic_config['coverSelector']}")?.getAttribute("{comic_config['coverAttr']}") || ""
            let author = document.querySelector("{comic_config['authorSelector']}")?.text || ""
            let description = document.querySelector("{comic_config['descSelector']}")?.text || ""
            let chapters = []
            document.querySelectorAll("{comic_config['chapterSelector']}").forEach(el => {{
                chapters.push(new Chapter({{
                    id: el.getAttribute("{comic_config['chapterIdAttr']}"),
                    title: el.querySelector("{comic_config['chapterTitleSelector']}")?.text || "",
                }}))
            }})
            return new ComicDetails({{
                id: id,
                title: title,
                cover: cover,
                author: author,
                description: description,
                chapters: chapters.reverse()
            }})
        }},
        loadEp: async (comicId, epId) => {{
            let url = "{comic_config['epUrl']}".replace("{{comicId}}", comicId).replace("{{epId}}", epId)
            let res = await Network.get(url)
            if (res.status !== 200) throw "请求失败"
            let document = new HtmlDocument(res.body)
            let images = []
            document.querySelectorAll("{comic_config['imageSelector']}").forEach(el => {{
                let src = el.getAttribute("{comic_config['imageAttr']}")
                if (src) images.push(src)
            }})
            return {{ images: images }}
        }}
    }}\n'''
        
        return source_code.replace('{{COMIC_SECTION}}', comic_code)
    
    def show_report(self):
        print('\n' + '='*60)
        print('          漫画源生成报告')
        print('='*60)
        
        for item in self.report:
            prefix = '  '
            if item['type'] == 'success':
                prefix = '  ✅ '
            elif item['type'] == 'error':
                prefix = '  ❌ '
            elif item['type'] == 'warning':
                prefix = '  ⚠️ '
            elif item['type'] == 'info':
                prefix = '  ℹ️ '
            print(f'{prefix}{item["message"]}')
        
        print('='*60)
        if hasattr(self, 'output_path'):
            print(f'\n生成的源文件: {self.output_path}')
    
    async def build(self):
        try:
            await self.check_dependencies()
            await self.run_tests()
            self.generate_source()
            self.show_report()
            return {'success': True, 'output_path': self.output_path}
        except Exception as error:
            self.log('error', f'构建失败: {error}')
            self.show_report()
            raise